src/main/java/com/liveramp/cascading_ext/yarn/YarnApiHelper.java

```diff
@@ -76,14 +79,62 @@ private static String GETRequest(String urlString) throws IOException {
   }
 
   public static Optional<ApplicationInfo> getYarnAppInfo(Configuration conf, String appId) {
-    String yarnApiAddress = conf.get("yarn.resourcemanager.webapp.address", "");
-    if ( !yarnApiAddress.isEmpty()) {
+    String yarnApiAddress = getYarnApiAddress(conf);
+    if (!yarnApiAddress.isEmpty()) {
       return getYarnAppInfo(yarnApiAddress, appId);
     } else {
       return Optional.empty();
     }
   }
 
+  /***
+   * Gets a yarn api address. This also works if you are using more than one resource manager.
+   * If there's more than one resource manager this method will return the address of the first
+   * resource manager with which it was able to successfully connect to. It will return an empty
+   * string if there were no resource manager it could successfully connect to.
+   */
+  private static String getYarnApiAddress(Configuration conf) {
+    Set<String> yarnApiAddresses = getYarnApiAddresses(conf);
+
+    try {
+      for (String yarnApiAddress : yarnApiAddresses) {
+        if (successfulConnection(yarnApiAddress)) {
+          return yarnApiAddress;
+        }
+      }
+    } catch (IOException e) {
+      LOG.error("Error getting yarn api address:", e);
+    }
+
+    return "";
+  }
+
+  static Set<String> getYarnApiAddresses(Configuration conf) {
+    String emptyString = "";
+    String rmIdsConf = conf.get("yarn.resourcemanager.ha.rm-ids", emptyString);
+
+    if (rmIdsConf.isEmpty()) {
+      String yarnApiAddress = conf.get("yarn.resourcemanager.webapp.address", emptyString);
+      return yarnApiAddress.isEmpty() ? Sets.newHashSet() : Sets.newHashSet(yarnApiAddress);
+    }
+
+    String[] allRmIds = rmIdsConf.split(",");
+    return Arrays.stream(allRmIds)
+        .map(rmId -> conf.get("yarn.resourcemanager.webapp.address." + rmId, emptyString))
+        .filter(address -> !address.isEmpty())
+        .collect(Collectors.toSet());
+  }
```

Comment:
Wondering about the `emptyString` here.

Efficiency: String constants are all interned and immutable so all references to `""` are the same thing anyway.

Readability: `""` potentially more readable than `emptyString`, but the interface for `Configuration` is that `get(key)` returns a `null` anyway which is sort of what you want in this case. Substituting with the `""` isn't really gaining anything right? You can `filter` with `Objects::notNull` which seems superior to supplying a default and then asking yourself if you have the default. I suppose it depends on how `null`-averse you are, or if you think that that the `conf` could legitimately contain empty values. Hmm. That last one sounds like of those Raymond Chen blog posts where one thing is producing weird stuff so the consuming side has workarounds whereas it's optimal to switch the producing side.

Reply:
At first, I thought there was a reason for using the empty string as the default value. After looking with more detail I couldn't find any real reason so I got rid of it. Thanks, Roshan.